StripeApplePay/StripeApplePay/Source/ApplePayContext/STPApplePayContext.swift

```diff
@@ -536,121 +557,73 @@ public class STPApplePayContext: NSObject, PKPaymentAuthorizationControllerDeleg
         with payment: PKPayment,
         completion: @escaping (PKPaymentAuthorizationStatus, Swift.Error?) -> Void
     ) {
-        // Helper to handle annoying logic around "Do I call completion block or dismiss + call delegate?"
-        let handleFinalState: ((PaymentState, Swift.Error?, StripeCore.Analytic) -> Void) = { state, error, analytic in
-            switch state {
-            case .error:
-                self.paymentState = .error
-                self.error = error
-                self.analyticsClient.log(analytic: analytic, apiClient: self.apiClient)
-                if self.didCancelOrTimeoutWhilePending {
-                    self.authorizationController?.dismiss {
-                        DispatchQueue.main.async {
-                            self.callDidCompleteDelegate(status: .error, error: self.error)
-                            self._end()
-                        }
+        // Helpers to handle annoying logic around "Do I call completion block or dismiss + call delegate?"
+        // Helper 1: Handle failure
+        func handleFailure(error: Swift.Error?) {
+            self.error = error ?? NSError.stp_genericErrorOccurredError()
```

Comment:
Defaulting to an error here feels a little awkward, i looked at a couple spots of where we're calling this function and it seems like there's always a error available?

Reply:
Yeah I'm not sure how to cleanly get the non-optional error in all the guards like
```
                        guard case .success(let setupIntent) = result else {
                            handleFailure(error: result.error)
                            return
                        }
```

If I use a switch statement, I'd have to nest everything inside the `success` case. Any ideas? Otherwise the optional error might be the least bad option.

Really we should make this method **way** better by using async and throws here - that would solve this problem too.